### DISCUSSION TITLE
ESP32-C3 & SX1262: Wake ESP from Light Sleep on packet receive

### DISCUSSION THREAD

**Discussion by XJDHDR** [jgromes/RadioLib#1015] [Q&A] [answered]
I have made a LoRa receiver using a Heltec HT-CT62, which internally uses an ESP32-C3 and SX1262. DIO1 is connected directly to GPIO3 inside the module. I am currently trying to have the ESP32 go into Light Sleep mode while waiting for a packet to arrive (SX1262 will remain awake, otherwise it won't receive anything). 

I tried `setDioIrqParams(0x0004, 0x0004)` to set DIO1 high when a packet is received and `gpio_wakeup_enable(GPIO_NUM_3, GPIO_INTR_HIGH_LEVEL); esp_sleep_enable_gpio_wakeup();` to make the ESP32 wake up when this happens. However, when I make the ESP sleep, it doesn't wake up no matter how many packets I transmit. I also tried `setDioIrqParams(0x001e, 0x001e)` instead, but this made no difference.

I tried probing GPIO3 with an oscilloscope while constantly sending the receiver packets, and saw a series of pulses. When I used the `0x0004` bitmask, each pulse was ~3.2V and appeared for ~550us, and there was ~120ms between each pulse. With the `0x001e` bitmask, a pair of pulses was created, with the two separated by ~22ms and the 2nd pulse in a pair was separated from the first in the next pair by ~100ms.

My current thoughts about this is that these pulses generated by the SX1262 are too short for the ESP to detect. Does this seem correct? Either way, does anyone have any suggestions for how I can make these interrupts wake the ESP (assuming it's possible)? 
Thanks!

**Comment by jgromes** [comment]
I don't know what are the limitations on ESP32 GPIO wakeup interrupts, but this: `gpio_wakeup_enable(GPIO_NUM_3, GPIO_INTR_HIGH_LEVEL);` seems to suggest you set up the interrupt for high level - why not a rising edge? Also, I would probably use `RADIOLIB_SX126X_IRQ_RX_DONE` as the interrupt source on SX126x, instead of the weird combinations you seem to be using.

**Reply by XJDHDR** [reply]
Thanks for responding.

> this: gpio_wakeup_enable(GPIO_NUM_3, GPIO_INTR_HIGH_LEVEL); seems to suggest you set up the interrupt for high level - why not a rising edge?

That is correct, it's set for high level. Unfortunately, I tried but edge detection doesn't work. Espressif's documentation [outright says](https://docs.espressif.com/projects/esp-idf/en/stable/esp32c3/api-reference/peripherals/gpio.html#_CPPv418gpio_wakeup_enable10gpio_num_t15gpio_int_type_t) that only low and high level are supported. I tried passing `GPIO_INTR_ANY_EDGE` and `GPIO_INTR_POS_EDGE` anyway, but the function returned `ESP_ERR_INVALID_ARG`. I also found [this comment](https://github.com/espressif/esp-idf/issues/12475#issuecomment-1784382747) which pretty much says that edge detection requires hardware that is powered down during sleep.

> Also, I would probably use RADIOLIB_SX126X_IRQ_RX_DONE as the interrupt source on SX126x, instead of the weird combinations you seem to be using.

I didn't know about those defines, and they would definitely make my code more readable. `0x0004` equates to `RADIOLIB_SX126X_IRQ_PREAMBLE_DETECTED`, and `0x001e` is `...IRQ_RX_DONE`, `...PREAMBLE_DETECTED`, `...SYNC_WORD_VALID` and `...HEADER_VALID`. I am using Implicit Headers so technically only `RX_DONE` and `SYNC_WORD_VALID` would actually trigger anything, and the other two are redundant.

**Reply by jgromes** [reply]
> Espressif's documentation [outright says](https://docs.espressif.com/projects/esp-idf/en/stable/esp32c3/api-reference/peripherals/gpio.html#_CPPv418gpio_wakeup_enable10gpio_num_t15gpio_int_type_t) that only low and high level are supported

Now that's something - thanks for pointing that out. Like I said, I don't know what the limitations are, but this seems like a major one. Perhaps there are other sleep modes of the ESP32 that could be used? I'm not sure much can be done from the SX126x side unforuntately.

**Reply by XJDHDR** [reply]
You're welcome. Unfortunately, the only other sleep mode is Deep Sleep, which saves more power, but pretty much resets the ESP and has more restrictions on what can wake it (only RTC domain GPIOs and Timer interrupts).

I guess the main question I have now is whether these pulse interrupts are how the SX1262 works, or if RadioLib is doing this (I strongly suspect the former)?

I'm currently looking into the possibility of using Timer interrupts to periodically wake the ESP and check the SX for new packets. As for making GPIO Wakeup work, I think it looks like a Latching Circuit would be required, where a DIO pulse would latch it on, then the ESP would unlatch it once it has woken up.

**Reply by jgromes** [reply]
> I guess the main question I have now is whether these pulse interrupts are how the SX1262 works,

I think the DIO1 line should remain raised until all associated interrupts are cleared. At least that's what the SX1262 datasheet says:

![Screenshot_82](https://github.com/jgromes/RadioLib/assets/3407192/377f22d2-b0f3-451f-9271-5782af652564)

So if the DIO line goes down, then either the interrupt was cleared, the SX1262 was reset or the ESP32 is pulling the pin low - in any case are you certain the ESP32 is not waking up?

**Reply by XJDHDR** [reply]
Strange thing happened. I wrote some firmware that consisted of only the below code, just to see if the basics work. When I did so, the DIO1 line stayed high (as you said it should), and the ESP32 woke from it's light sleep. I have no idea why this didn't work the previous times I tried it. All I can think of is either something in the previous code that didn't work correctly, or maybe PlatformIO didn't write it correctly, or something else.

Anyway, the next thing I need to work out is if I can still have the interrupt based method of receiving packets work with the ESP waking from sleep.

```
#include <Arduino.h>
#include <RadioLib.h>
#include <SPI.h>
#include <stdint.h>
#include <thread>

#define LoRa_nss 8
#define LoRa_dio1 GPIO_NUM_3
#define LoRa_nrst 5
#define LoRa_busy 4

#define LoRa_MOSI 7
#define LoRa_MISO 6
#define LoRa_SCK 10

#define PACKET_LENGTH 2

const uint16_t irqMask = RADIOLIB_SX126X_IRQ_RX_DONE;

SX1262 sx1262 = new Module(LoRa_nss, LoRa_dio1, LoRa_nrst, LoRa_busy);

IRAM_ATTR void onPacketReceived()
{
}

void setupRadio()
{
	int state;
	state = sx1262.begin(921.5, 125.0, 9, 7, 0x12, 22);
	if (state != RADIOLIB_ERR_NONE)
	{
		Serial.printf("Failed to initialise radio with error: %i", state);
		Serial.println();
	}
	else
	{
		sx1262.implicitHeader(PACKET_LENGTH);

		sx1262.setPacketReceivedAction(onPacketReceived);

		state = sx1262.startReceive(RADIOLIB_SX126X_RX_TIMEOUT_INF, irqMask, irqMask);
		if (state != RADIOLIB_ERR_NONE)
		{
			Serial.printf("Failed to start listening for packets with error: %i", state);
			Serial.println();
		}

		sx1262.clearIrqStatus(irqMask);
	}
}

void setup()
{
	Serial.begin(115200);

	// Sleep for a bit to give terminals a chance to connect.
	std::this_thread::sleep_for(std::chrono::milliseconds(5000));

	Serial.println("Setup started.");

	SPI.begin(LoRa_SCK, LoRa_MISO, LoRa_MOSI);

	setupRadio();

	gpio_wakeup_enable(LoRa_dio1, GPIO_INTR_HIGH_LEVEL);
	esp_sleep_enable_gpio_wakeup();

	Serial.println("About to go to sleep.");
	esp_light_sleep_start();
//	sx1262.clearIrqStatus(irqMask);
	Serial.println("Sleep ended.");

	Serial.println("Setup finished.");
}

void loop()
{
	Serial.println("New loop started.");
	std::this_thread::sleep_for(std::chrono::milliseconds(5000));

//	if (digitalRead(LoRa_dio1) == HIGH)
//	{
//		sx1262.clearIrqStatus(irqMask);
//	}
}
```

**Reply by jgromes** [reply]
Glad you were able to get it to work!

> the next thing I need to work out is if I can still have the interrupt based method of receiving packets

I think you should be able to. One note about your code, you can pass IRQ flags to `startReceive`, e.g.:

`radio.startReceive(RADIOLIB_SX126X_RX_TIMEOUT_INF, RADIOLIB_SX126X_IRQ_RX_DONE, RADIOLIB_SX126X_IRQ_RX_DONE);`

And you don't have to change DIO1 to input, that's done by the library.

**Reply by XJDHDR** [reply]
Thanks, I've updated my last post.

Just to give an update, it doesn't appear to be possible to use RadioLib's Interrupt method with Light Sleep. I've found with testing that you have to call the `startReceive()` function in addition to setting the desired IRQ Mask in order to make DIO1 go high. Merely calling `setDioIrqParams(<Mask>)` is not enough (I'm assuming `setRx()` puts the SX in a "receive" state or something). 

However, calling `setPacketReceivedAction()` causes the ESP to enter a "limbo" state where it is awake, but stops executing any more of your firmware. It also starts drawing 45mA in this state, which is suspiciously the same current draw that happens when you're writing firmware to the ESP.

So at the moment, I'm not using the `setPacketReceivedAction()` function, but this does mean that I need a different method of processing the packets' payloads. I'm looking into using `readBuffer()` to get that data.

**Reply by jgromes** [reply]
> you have to call the startReceive()

`startReceive` is what will ... well, start receiving. The IRQ methods only set which interrupts are enabled and propagated to the DIO line.

> setPacketReceivedAction() causes the ESP to enter a "limbo" state

Looking at your code, it looks like you're missing `IRAM_ATTR` for your interrupt service routine `onPacketReceived`. On ESP32 when using Arduino, this is required.

**Reply by XJDHDR** [reply, marked as answer]
> it looks like you're missing IRAM_ATTR for your interrupt service routine onPacketReceived. On ESP32 when using Arduino, this is required.

That was indeed the problem, and I've amended my previous code accordingly. Added the Attribute, and the ESP now properly wakes up. I didn't know it was required, and this hasn't caused issues before. I guess the function can't be read from flash while/right after waking up.

Otherwise, I think this should be everything. Thanks for the help. 
So to answer my original question for anyone else who reads this, I think the best thing to do if you are having issues with sleep is to flash the code in [my previous post](https://github.com/jgromes/RadioLib/discussions/1015#discussioncomment-8873435) and see how things go. If your ESP doesn't wake, that would indicate a hardware problem. If it does, there's something wrong with your firmware.

And on that subject, I also found why my firmware wasn't waking the ESP. In my original (not working) code, I had the `gpio_wakeup_enable(GPIO_NUM_3, GPIO_INTR_HIGH_LEVEL); esp_sleep_enable_gpio_wakeup();` functions in the Constructor for a class that is instantiated pretty much as soon as possible. This appears to make the functions not properly run because when I moved those two function calls into the Arduino `Setup()` function, my code now also correctly wakes the ESP.
